core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java

```diff
@@ -76,14 +81,34 @@ public Boolean evaluate(ProjectConfig config, Map<String, ?> attributes) {
         Object userAttributeValue = attributes.get(name);
 
         if (!"custom_attribute".equals(type)) {
-            MatchType.logger.error(String.format("condition type not equal to `custom_attribute` %s", type));
+            logger.error("Audience condition \"{}\" has an unknown condition type: {}", this.toString(), type);
             return null; // unknown type
         }
         // check user attribute value is equal
         try {
-            return MatchType.getMatchType(match, value).getMatcher().eval(userAttributeValue);
+            Match matchType = MatchType.getMatchType(match, value).getMatcher();
+            Boolean result = matchType.eval(userAttributeValue);
+
+            if (!(matchType instanceof NullMatch) && result == null) {
+                if (!attributes.containsKey(name)) {
+                    //Missing attribute value
+                    logger.debug("Audience condition \"{}\" evaluated as UNKNOWN because no user attribute for \"{}\" key is passed", this.toString(), name);
+                } else {
+                    //if attribute value is not valid
+                    logger.warn(
+                        "Audience condition \"{}\" evaluated as UNKNOWN because the value for user attribute \"{}\" is inapplicable: \"{}\"",
+                        this.toString(),
+                        name,
+                        userAttributeValue);
+                }
+            } else if (matchType instanceof NullMatch) {
+                logger.error("Audience condition \"{}\" " + ((NullMatch) matchType).getMatchTypeError().toString(),
```

Comment:
In MatchType.java and MatchTypeError.java, I'm suggesting that we reduce the number of entries in the `MatchTypeError` enum from 2 to 1. Maybe we can get away with inlining that 1 message here instead of defining it in an enum? I'll leave it up to you all.

Reply:
[Explained in MatchType.java](https://github.com/optimizely/java-sdk/pull/251#discussion_r247898163)